### DISCUSSION TITLE
Date and Time update independent from HA

### DISCUSSION THREAD

**Discussion by denes44** [Blackymas/NSPanel_HA_Blueprint#735] [Ideas]
Another great fallback feature would be, if the date and time updates are not dependent on the Home Assistant connection.
This is the code I used for my personal HMI project:
```
time:
  - platform: homeassistant
    id: homeassistant_time
    on_time:
      - seconds: 0
        minutes: /1
        then:
          - lambda: id(disp1).set_component_text_printf("Home.time", "%02i:%02i", id(homeassistant_time).now().hour, id(homeassistant_time).now().minute);
          - lambda: id(disp1).set_component_text_printf("Home.date", "%i-%02i-%02i", id(homeassistant_time).now().year, id(homeassistant_time).now().month, id(homeassistant_time).now().day_of_month);
    on_time_sync:
        then:
          - wait_until:
              switch.is_on: nextion_init
          - lambda: id(disp1).set_component_text_printf("Home.time", "%02i:%02i", id(homeassistant_time).now().hour, id(homeassistant_time).now().minute);
          - lambda: id(disp1).set_component_text_printf("Home.date", "%i-%02i-%02i", id(homeassistant_time).now().year, id(homeassistant_time).now().month, id(homeassistant_time).now().day_of_month);
```
This time platform does the time synchronizing part between the HA and esphome, and the esphome could update the time on the display without a connection to HA. So if somebody has a problem with HA for a time, not only the buttons on the panel would work, but also the time would not stop.
For this to work, [#733](https://github.com/Blackymas/NSPanel_HA_Blueprint/discussions/733) will be needed (global variables for the home page).

So for this to work, we need to send and store the time and date string selected on the blueprint on the ESP device, and after that, it can do the time updating itself.
This can maybe also support displaying the seconds, if somebody needs this feature. I'm not sure however how sending the time every second on the serial port will affect the other communications for the display.
But that would not be the biggest advantage, but the fact, that the date/time display would be independent from HA, and less work from the blueprint/automation/wifi connection/etc.

**Comment by edwardtfn** [comment]
I really like the idea of moving the date and time to ESPHome, but this will be a challenge to handle translations (date only).
I don't know exactly how ESPHome handles localization, but in HA it happens on frontend only, so if we use date format to get weekdays names or month names it will always return in English. I believe this was the main reason for handling dates on the Blueprint from the beginning.

I mean, if you you have your system in Portuguese, and use `{{ now() | as_timestamp | timestamp_custom("%A, %b %-d") }}` it will return `Thursday, Apr 27` rather than `Segunda-feira, Abr 27`. The language used to generate that by Home Assistant will always be English regardless user's selected language.
Time would be easy, as just numbers are shown.

**Reply by denes44** [reply]
Yes, translation would be an issue. Well not in issue, it's just not trivial.
Locale is not working in esphome for strftime, so that is not an option.

Regarding date I think we have two options:
- Leave it in the blueprint/automation
- Do the translation work to esphome

Only the latter would achieve the independent functioning from HA. Other project would probably benefit from a date translating code for esphome, so I believe it's worth doing. This project already has the translations, so we only need to do the code.

**Reply by edwardtfn** [reply]
> Yes, translation would be an issue. Well not in issue, it's just not trivial.

Not trivial, but not difficult.
We can create a global variable in ESPHome containing an array of strings and a service to update that string with the translations. We can even extend this to short week days ("Mon", "Tue", ...) and months ("Jan", "Feb", ...) if we want implement this in the future.
Something like this (not tested):
```yaml
globals:
  - id: weekdays
    type: std::vector<std::string>
    restore_value: true
    initial_value: '{"Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"}'
```
The Blueprint would call that service within a few triggers (HA starts, automation reloaded, panel boot(?), etc.), so this wouldn't overload the comms.
The concerns about adding this to ESPHome side is always on its limited memory, but I don't think a solution like this will have a significant impact.

**Reply by edwardtfn** [reply]
But this will also require implementing on the ESPHome side a string replacement for the translations, similar to what was done in the Blueprint on #744.

**Comment by denes44** [comment]
Another thing related to the fallback functions is restarting the panel and HA is not available.
For example if after a power outage the HA fails to start for some reason, the panel would stuck at the boot page, right?
Maybe if we do this datetime independent thing, it's worth changing that the panel boots to the home page, displays the home page with the clock, and only change to the boot screen after esphome get a connection to HA (`api.connected`).

**Comment by edwardtfn** [comment]
@denes44, I will try to create a working prototype for this.
Do you wanna join me on this journey?

I will be using this branch: https://github.com/Blackymas/NSPanel_HA_Blueprint/tree/poc-datetime-on-esphome

**Reply by denes44** [reply]
Yes, sure.
I'm more familiar with the esphome side, I only started to familiar with the templates.

**Reply by edwardtfn** [reply]
You mentioned about changes needed in the Nextion side... If you can start with that it would be super nice.

**Reply by denes44** [reply]
Well, the problem is, you can't really merge the Nextion project file. There is an open PR which modified the HMI files (#740), and if I modify the HMI file on your branch again, it would be almost impossible to merge the two separately edited files. But the nextion modification is very easy, can be done in 5 minutes, and it is not needed for the testing, because if you are on the homepage it does not matter if the variables on the home page are global or local.

**Reply by denes44** [reply]
> I will be using this branch: https://github.com/Blackymas/NSPanel_HA_Blueprint/tree/poc-datetime-on-esphome

I started doing the esphome part, but I don't have write access here.

**Reply by edwardtfn** [reply]
Mmmm
Can you create a PR there?

**Reply by edwardtfn** [reply]
I was focusing on the release the past days and now I got sick, but I want jump on this as soon as possible.

**Reply by edwardtfn** [reply]
Anyway, we can wait a couple of days more to make sure nothing critical comes from the release of v3.3 and then let's look at your PRS at `dev`, so we can continue working from there, directly on `dev`.

**Comment by rleongcs** [comment]
While you guys are on this...any possibility to incorporate the below? Or something similar so we can know if it is connected or not.

```
  - platform: status
    name: "Status"
```

**Reply by denes44** [reply]
This is my common_stuff.yaml config for every esphome device I use:
```
binary_sensor:
  - platform: status
    name: ${devicename} Status

button:
  - platform: restart
    name: ${devicename} Restart
    id: _restart
  - platform: safe_mode
    name: ${devicename} Restart (Safe Mode)
    id: _restart_safe
  - platform: shutdown
    name: ${devicename} Shutdown
    id: _shutdown
    
sensor:
  - platform: wifi_signal
    name: ${devicename} Wifi
    id: wifi_signal_strength
    update_interval: 60s
    filters:
    - sliding_window_moving_average:
        window_size: 15
        send_every: 15
```

You can add the status platform to you yaml config file, which includes the nspanel_esphome.yaml file, but I see no reason not to include this in the base code.

**Comment by edwardtfn** [comment]
Time update was included on v4.0: #1057 
Date update still needing HA. I've the code almost ready, but it takes quite a lot of memory to transfer all the strings with months and week days, so I'm not sure it worth when compared to the benefit.
Please let me know your thoughts.
